libs/contrib/Data/String/Parser/Expression.idr

```diff
@@ -0,0 +1,125 @@
+-- Port of https://hackage.haskell.org/package/parsec-3.1.13.0/docs/Text-Parsec-Expr.html
+-- Original License	BSD-style: https://hackage.haskell.org/package/parsec-3.1.13.0/src/LICENSE
+
+module Data.String.Parser.Expression
+
+import Control.Monad.Identity
+import Data.String.Parser
+
+public export
+data Assoc = AssocNone
+           | AssocLeft
+           | AssocRight
+
+public export
+data Operator a = Infix (Parser (a -> a -> a)) Assoc
+                | Prefix (Parser (a -> a))
+                | Postfix (Parser (a -> a))
+
+public export
+OperatorTable : Type -> Type
+OperatorTable a = List (List (Operator a))
+
+public export
+BinaryOperator : Type -> Type
+BinaryOperator a = List (Parser (a -> a -> a))
+
+public export
+UnaryOperator : Type -> Type
+UnaryOperator a = List (Parser (a -> a))
+
+public export
+data Ops a = BinOp (BinaryOperator a) | UnOp (UnaryOperator a)
+
+public export
+ReturnType : Type -> Type
+ReturnType a = (BinaryOperator a, BinaryOperator a, BinaryOperator a, UnaryOperator a, UnaryOperator a)
+
+toParserBin : BinaryOperator a -> Parser (a -> a -> a)
+toParserBin [] = fail "couldn't create binary parser"
+toParserBin (x :: xs) = x <|> toParserBin xs
+
+toParserUn : UnaryOperator a -> Parser (a -> a)
+toParserUn [] = fail "couldn't create unary parser"
+toParserUn (x :: xs) = x <|> toParserUn xs
+
+ambigious : (assoc : String) -> (op : Parser (a -> a -> a)) -> Parser a
+ambigious assoc op = do op
+                        fail ("ambiguous use of a " ++ assoc ++ " associative operator")
+
+mutual
+  mkRassocP : (amLeft : Parser a) -> (amNon : Parser a) -> (rassocOp : Parser (a -> a -> a)) -> (termP : Parser a) -> (x : a) -> Parser a
+  mkRassocP amLeft amNon rassocOp termP x =
+   (do f <- rassocOp
+       y <- do z <- termP ; mkRassocP1 amLeft amNon rassocOp termP z
+       pure (f x y))
+   <|> (amLeft)
+   <|> (amNon)
+
+  mkRassocP1 : (amLeft : Parser a) -> (amNon : Parser a) -> (rassocOp : Parser (a -> a -> a)) -> (termP : Parser a) -> (x : a) -> Parser a
+  mkRassocP1 amLeft amNon rassocOp termP x = (mkRassocP amLeft amNon rassocOp termP x) <|> pure x
+
+mutual
+  mkLassocP : (amRight : Parser a) -> (amNon : Parser a) -> (lassocOp : Parser (a -> a -> a)) -> (termP : Parser a) -> (x : a) -> Parser a
+  mkLassocP amRight amNon lassocOp termP x =
+    (do f <- lassocOp
+        y <- termP
+        mkLassocP1 amRight amNon lassocOp termP (f x y))
+    <|> amRight
+    <|> amNon
+
+  mkLassocP1 : (amRight : Parser a) -> (amNon : Parser a) -> (lassocOp : Parser (a -> a -> a)) -> (termP : Parser a) -> (x : a) -> Parser a
+  mkLassocP1 amRight amNon lassocOp termP x = mkLassocP amRight amNon lassocOp termP x <|> pure x
```

Comment:
These have useful generalisations as combinators that could be moved to `Data.String.Parser`:

```
hchainl : Parser init -> Parser (init -> arg -> init) -> Parser arg -> Parser init
hchainr : Parser arg -> Parser (arg -> end -> end) -> Parser end -> Parser end
```

The `h` stands for "heterogeneous" as opposed to the usual homogeneous chain combinators.

Reply:
That definitely sounds cool, but I'm sure I have the skills to pull of a refactor like that 😅

I seem to have extra params (`Parser a`, and `a : Type`) than the type signatures you provided. Not sure I can square that difference...

Good to know that's what the `h` stands for!